query/src/frontend/influxrpc.rs

```diff
@@ -92,16 +153,263 @@ impl InfluxRPCPlanner {
+                        known_columns.append(&mut names);
+                    }
+                    None => {
+                        debug!(
+                            table_name = table_name.as_str(),
+                            chunk_id = chunk.id(),
+                            "column names need full plan"
+                        );
+                        // can't get columns only from metadata, need
+                        // a general purpose plan
+                        need_full_plans
+                            .entry(table_name)
+                            .or_insert_with(Vec::new)
+                            .push(chunk.clone());
+                    }
+                }
+            }
+        }
+
+        let mut builder = StringSetPlanBuilder::new();
+
+        // At this point, we have a set of column names we know pass
+        // in `known_columns`, and potentially some tables in chunks
+        // that we need to run a plan to know if they pass the
+        // predicate.
+        if !need_full_plans.is_empty() {
+            // TODO an additional optimization here would be to filter
+            // out chunks (and tables) where all columns in that chunk
+            // were already known to have data (based on the contents of known_columns)
+
+            for (table_name, chunks) in need_full_plans.into_iter() {
+                let plan = self
+                    .tag_column_names_plan(database, &table_name, &predicate, chunks)
+                    .await?;
+
+                if let Some(plan) = plan {
+                    builder = builder.append(plan)
+                }
+            }
+        }
+
+        // add the known columns we could find from metadata only
+        builder
+            .append(known_columns.into())
+            .build()
+            .context(CreatingStringSet)
+    }
+
+    /// removes any columns from Names that are not "Tag"s in the Influx Data
+    /// Model
+    fn restrict_to_tags(&self, schema: &Schema, names: BTreeSet<String>) -> BTreeSet<String> {
+        names
+            .into_iter()
+            .filter(|col_name| {
+                let idx = schema.find_index_of(col_name).unwrap();
+                matches!(schema.field(idx), (Some(InfluxColumnType::Tag), _))
+            })
+            .collect()
+    }
+
+    /// Creates a DataFusion LogicalPlan that returns column *names* as a
+    /// single column of Strings for a specific table
+    ///
+    /// The created plan looks like:
+    ///
+    ///  Extension(PivotSchema)
+    ///    Filter(predicate)
+    ///      TableScan (of chunks)
+    async fn tag_column_names_plan<D>(
+        &self,
+        //  I needed an unused parameter with type D so that Rust
+        //  would let me call this function
+        _database: &D,
+        table_name: &str,
+        predicate: &Predicate,
+        chunks: Vec<Arc<<D as Database>::Chunk>>,
+    ) -> Result<Option<StringSetPlan>>
+    where
+        D: Database + 'static,
+    {
```

Comment:
Because [`Database::Chunk` has to implement `PartitionChunk`](https://github.com/influxdata/influxdb_iox/blob/5ac4ee2f49141a872f0b1348d864b75489ccf2ea/query/src/lib.rs#L41), would it be sufficient here to say:

```suggestion
    async fn tag_column_names_plan<P>(
        &self,
        table_name: &str,
        predicate: &Predicate,
        chunks: Vec<Arc<P>>,
    ) -> Result<Option<StringSetPlan>>
    where
        P: PartitionChunk + 'static,
    {
```

? It compiles for me (when I remove the extra param at the call site too), but I'm not sure if this is correct?

Reply:
YES! Thank you @carols10cents  that seems so obvious now, but I am not sure why it wasn't in the heat of the coding. Much better. 🙏 Fixed in 386b9599